File: scripts/convert-pdf-to-images.js
#!/usr/bin/env node
const fs = require("fs");
const path = require("path");
const { convert } = require("pdf-poppler");
const sharp = require("sharp");

function getArgument(name, fallback) {
  const index = process.argv.indexOf(`--${name}`);
  if (index === -1 || index === process.argv.length - 1) {
    return fallback;
  }

  return process.argv[index + 1];
}

const volumeId = getArgument("volume", "volume1");
const defaultPdfFile =
  volumeId === "volume2" ? "shifa-shareef-v2.pdf" : "shifa-shareef-v1.pdf";
const pdfFileName = getArgument("pdf", defaultPdfFile);
const PDF_PATH = path.join(__dirname, "../assets/pdf", pdfFileName);
const OUTPUT_DIR = path.join(__dirname, "../assets/pages", volumeId);
const TEMP_DIR = path.join(__dirname, `../temp-pages-${volumeId}`);
const QUALITY = 85;
const MAX_WIDTH = 1080;

async function convertPdfToImages() {
  console.log("Converting PDF to images...");
  console.log(`Volume: ${volumeId}`);
  console.log(`PDF: ${PDF_PATH}`);
  console.log(`Output: ${OUTPUT_DIR}`);

  if (!fs.existsSync(PDF_PATH)) {
    console.error(`PDF file not found: ${PDF_PATH}`);
    process.exit(1);
  }

  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  fs.mkdirSync(TEMP_DIR, { recursive: true });

  try {
    await convert(PDF_PATH, {
      format: "png",
      out_dir: TEMP_DIR,
      out_prefix: "page",
      page: null,
      scale: 2048,
    });

    const pngFiles = fs
      .readdirSync(TEMP_DIR)
      .filter((file) => file.endsWith(".png"))
      .sort();

    let totalSize = 0;

    for (let index = 0; index < pngFiles.length; index += 1) {
      const pngFile = pngFiles[index];
      const pageNum = String(index + 1).padStart(3, "0");
      const inputPath = path.join(TEMP_DIR, pngFile);
      const outputPath = path.join(OUTPUT_DIR, `page-${pageNum}.webp`);
      const image = sharp(inputPath);
      const metadata = await image.metadata();

      if (metadata.width && metadata.width > MAX_WIDTH && metadata.height) {
        const ratio = MAX_WIDTH / metadata.width;
        image.resize(MAX_WIDTH, Math.round(metadata.height * ratio), {
          fit: "inside",
          withoutEnlargement: true,
        });
      }

      await image.webp({ quality: QUALITY, effort: 6 }).toFile(outputPath);
      totalSize += fs.statSync(outputPath).size / 1024;
      fs.unlinkSync(inputPath);
      console.log(`Page ${pageNum}/${pngFiles.length}`);
    }

    fs.rmdirSync(TEMP_DIR);

    console.log("Conversion complete.");
    console.log(`Pages: ${pngFiles.length}`);
    console.log(`Total size: ${(totalSize / 1024).toFixed(2)} MB`);
  } catch (error) {
    console.error("Conversion failed:", error.message);
    process.exit(1);
  }
}

convertPdfToImages();

File: scripts/convert-pdf-to-images.py
#!/usr/bin/env python3
"""
Convert PDF to optimized WebP images for smooth mobile reading experience.

Requirements:
    pip install pdf2image pillow

System dependencies:
    - poppler-utils (Linux/Mac: brew install poppler)
    - Windows: Download poppler binaries
"""

import os
from pathlib import Path
from pdf2image import convert_from_path
from PIL import Image

# Configuration
PDF_PATH = "../assets/pdf/shifa-shareef-v1.pdf"
OUTPUT_DIR = "../assets/pages"
IMAGE_FORMAT = "webp"
QUALITY = 85
DPI = 150  # Good balance between quality and file size
MAX_WIDTH = 1080  # Mobile screen width

def convert_pdf_to_images():
    """Convert PDF pages to optimized WebP images."""
    
    # Create output directory
    output_path = Path(__file__).parent / OUTPUT_DIR
    output_path.mkdir(parents=True, exist_ok=True)
    
    # Get PDF path
    pdf_path = Path(__file__).parent / PDF_PATH
    
    if not pdf_path.exists():
        print(f"❌ PDF not found: {pdf_path}")
        return
    
    print(f"📄 Converting PDF: {pdf_path}")
    print(f"📁 Output directory: {output_path}")
    print(f"⚙️  Settings: DPI={DPI}, Quality={QUALITY}%, Format={IMAGE_FORMAT.upper()}")
    print()
    
    try:
        # Convert PDF to images
        print("🔄 Converting PDF pages to images...")
        images = convert_from_path(
            pdf_path,
            dpi=DPI,
            fmt='png',  # Intermediate format
            thread_count=4  # Use multiple threads for speed
        )
        
        total_pages = len(images)
        print(f"✅ Extracted {total_pages} pages")
        print()
        
        # Process and save each page
        total_size = 0
        for i, image in enumerate(images, start=1):
            page_num = str(i).zfill(3)  # 001, 002, etc.
            output_file = output_path / f"page-{page_num}.{IMAGE_FORMAT}"
            
            # Resize if needed (maintain aspect ratio)
            if image.width > MAX_WIDTH:
                ratio = MAX_WIDTH / image.width
                new_height = int(image.height * ratio)
                image = image.resize((MAX_WIDTH, new_height), Image.Resampling.LANCZOS)
            
            # Save as WebP with optimization
            image.save(
                output_file,
                format=IMAGE_FORMAT.upper(),
                quality=QUALITY,
                method=6  # Maximum compression effort
            )
            
            file_size = output_file.stat().st_size / 1024  # KB
            total_size += file_size
            
            print(f"✓ Page {page_num}/{total_pages} → {file_size:.1f} KB")
        
        print()
        print(f"🎉 Conversion complete!")
        print(f"📊 Total size: {total_size / 1024:.2f} MB")
        print(f"📊 Average per page: {total_size / total_pages:.1f} KB")
        print(f"📁 Images saved to: {output_path}")
        
    except Exception as e:
        print(f"❌ Error: {e}")
        print()
        print("💡 Make sure you have installed:")
        print("   pip install pdf2image pillow")
        print("   brew install poppler  # Mac")
        print("   apt install poppler-utils  # Linux")

if __name__ == "__main__":
    convert_pdf_to_images()
